File: go/builder.go
package membuffers

type MessageWriter interface {
	HexDump(prefix string, offsetFromStart Offset) (err error)
	Write(buf []byte) (err error)
	GetSize() Offset
	CalcRequiredSize() Offset
}

type InternalBuilder struct {
	size     Offset
	building bool
}

func (w *InternalBuilder) Reset() {
	w.size = 0
}

// override me
func (w *InternalBuilder) Write(buf []byte) (err error) {
	defer func() {
		if r := recover(); r != nil {
			err = &ErrBufferOverrun{}
		}
	}()
	w.Reset()
	return nil
}

// instead of implementing Write with the actual fields, this allows to take a ready-made raw buffer
func (w *InternalBuilder) WriteOverrideWithRawBuffer(buf []byte, overrideWithRawBuffer []byte) error {
	if buf != nil {
		if len(buf) < len(overrideWithRawBuffer) {
			return &ErrBufferOverrun{}
		}
		copy(buf, overrideWithRawBuffer)
	}
	w.size = Offset(len(overrideWithRawBuffer))
	return nil
}

func (w *InternalBuilder) CalcRequiredSize() Offset {
	w.Write(nil)
	return w.GetSize()
}

func (w *InternalBuilder) GetSize() Offset {
	return w.size
}

func (w *InternalBuilder) NotifyBuildStart() {
	if w.building {
		panic("concurrent build: builder is not thread safe (started building from multiple goroutines)")
	}
	w.building = true // not a mutex for performance reasons
}

func (w *InternalBuilder) NotifyBuildEnd() {
	w.building = false
}

func (w *InternalBuilder) WriteUint8(buf []byte, v uint8) {
	w.size = alignOffsetToType(w.size, TypeUint8)
	if buf != nil {
		WriteUint8(buf[w.size:], v)
	}
	w.size += FieldSizes[TypeUint8]
}

func (w *InternalBuilder) WriteUint16(buf []byte, v uint16) {
	w.size = alignOffsetToType(w.size, TypeUint16)
	if buf != nil {
		WriteUint16(buf[w.size:], v)
	}
	w.size += FieldSizes[TypeUint16]
}

func (w *InternalBuilder) WriteUint32(buf []byte, v uint32) {
	w.size = alignOffsetToType(w.size, TypeUint32)
	if buf != nil {
		WriteUint32(buf[w.size:], v)
	}
	w.size += FieldSizes[TypeUint32]
}

func (w *InternalBuilder) WriteUint64(buf []byte, v uint64) {
	w.size = alignOffsetToType(w.size, TypeUint64)
	if buf != nil {
		WriteUint64(buf[w.size:], v)
	}
	w.size += FieldSizes[TypeUint64]
}

func (w *InternalBuilder) WriteBytes(buf []byte, v []byte) {
	w.size = alignOffsetToType(w.size, TypeBytes)
	if buf != nil {
		WriteOffset(buf[w.size:], Offset(len(v)))
	}
	w.size += FieldSizes[TypeBytes]
	w.size = alignDynamicFieldContentOffset(w.size, TypeBytes)
	if v != nil {
		if buf != nil {
			copy(buf[w.size:], v)
		}
		w.size += Offset(len(v))
	}
}

func (w *InternalBuilder) WriteString(buf []byte, v string) {
	w.WriteBytes(buf, []byte(v))
}

func (w *InternalBuilder) WriteUnionIndex(buf []byte, unionIndex uint16) {
	w.size = alignOffsetToType(w.size, TypeUnion)
	if buf != nil {
		WriteUnionType(buf[w.size:], unionIndex)
	}
	w.size += FieldSizes[TypeUnion]
}

func (w *InternalBuilder) WriteUint8Array(buf []byte, v []uint8) {
	w.size = alignOffsetToType(w.size, TypeUint8Array)
	if buf != nil {
		WriteOffset(buf[w.size:], Offset(len(v))*FieldSizes[TypeUint8])
	}
	w.size += FieldSizes[TypeUint8Array]
	w.size = alignDynamicFieldContentOffset(w.size, TypeUint8Array)
	for _, vv := range v {
		w.WriteUint8(buf, vv)
	}
}

func (w *InternalBuilder) WriteUint16Array(buf []byte, v []uint16) {
	w.size = alignOffsetToType(w.size, TypeUint16Array)
	if buf != nil {
		WriteOffset(buf[w.size:], Offset(len(v))*FieldSizes[TypeUint16])
	}
	w.size += FieldSizes[TypeUint16Array]
	w.size = alignDynamicFieldContentOffset(w.size, TypeUint16Array)
	for _, vv := range v {
		w.WriteUint16(buf, vv)
	}
}

func (w *InternalBuilder) WriteUint32Array(buf []byte, v []uint32) {
	w.size = alignOffsetToType(w.size, TypeUint32Array)
	if buf != nil {
		WriteOffset(buf[w.size:], Offset(len(v))*FieldSizes[TypeUint32])
	}
	w.size += FieldSizes[TypeUint32Array]
	w.size = alignDynamicFieldContentOffset(w.size, TypeUint32Array)
	for _, vv := range v {
		w.WriteUint32(buf, vv)
	}
}

func (w *InternalBuilder) WriteUint64Array(buf []byte, v []uint64) {
	w.size = alignOffsetToType(w.size, TypeUint64Array)
	if buf != nil {
		WriteOffset(buf[w.size:], Offset(len(v))*FieldSizes[TypeUint64])
	}
	w.size += FieldSizes[TypeUint64Array]
	w.size = alignDynamicFieldContentOffset(w.size, TypeUint64Array)
	for _, vv := range v {
		w.WriteUint64(buf, vv)
	}
}

func (w *InternalBuilder) WriteBytesArray(buf []byte, v [][]byte) {
	w.size = alignOffsetToType(w.size, TypeBytesArray)
	sizePlaceholderOffset := w.size
	w.size += FieldSizes[TypeBytesArray]
	w.size = alignDynamicFieldContentOffset(w.size, TypeBytesArray)
	contentSizeStartOffset := w.size
	for _, vv := range v {
		w.WriteBytes(buf, vv)
	}
	contentSize := w.size - contentSizeStartOffset
	if buf != nil {
		WriteOffset(buf[sizePlaceholderOffset:], contentSize)
	}
}

func (w *InternalBuilder) WriteStringArray(buf []byte, v []string) {
	w.size = alignOffsetToType(w.size, TypeStringArray)
	sizePlaceholderOffset := w.size
	w.size += FieldSizes[TypeStringArray]
	w.size = alignDynamicFieldContentOffset(w.size, TypeStringArray)
	contentSizeStartOffset := w.size
	for _, vv := range v {
		w.WriteString(buf, vv)
	}
	contentSize := w.size - contentSizeStartOffset
	if buf != nil {
		WriteOffset(buf[sizePlaceholderOffset:], contentSize)
	}
}

func (w *InternalBuilder) WriteMessage(buf []byte, v MessageWriter) (err error) {
	w.size = alignOffsetToType(w.size, TypeMessage)
	sizePlaceholderOffset := w.size
	w.size += FieldSizes[TypeMessage]
	w.size = alignDynamicFieldContentOffset(w.size, TypeMessage)
	if buf != nil {
		err = v.Write(buf[w.size:])
	} else {
		err = v.Write(nil)
	}
	contentSize := v.GetSize()
	w.size += contentSize
	if buf != nil {
		WriteOffset(buf[sizePlaceholderOffset:], contentSize)
	}
	return
}

func (w *InternalBuilder) WriteMessageArray(buf []byte, v []MessageWriter) (err error) {
	w.size = alignOffsetToType(w.size, TypeMessageArray)
	sizePlaceholderOffset := w.size
	w.size += FieldSizes[TypeMessageArray]
	w.size = alignDynamicFieldContentOffset(w.size, TypeMessageArray)
	contentSizeStartOffset := w.size
	for _, vv := range v {
		err = w.WriteMessage(buf, vv)
		if err != nil {
			return
		}
	}
	contentSize := w.size - contentSizeStartOffset
	if buf != nil {
		WriteOffset(buf[sizePlaceholderOffset:], contentSize)
	}
	return nil
}

File: javascript/builder.js
import {alignOffsetToType, alignDynamicFieldContentOffset} from './message';
import {FieldTypes, FieldSizes} from './types';
import {getTextEncoder} from './text';

export class InternalBuilder {

  constructor() {
    this.size = 0;
  }

  reset() {
    this.size = 0;
  }

  calcRequiredSize() {
    this.write(null);
    return this.getSize();
  }

  getSize() {
    return this.size;
  }

  writeUint8(buf, v) {
    this.size = alignOffsetToType(this.size, FieldTypes.TypeUint8);
    if (buf) {
      new DataView(buf.buffer, buf.byteOffset).setUint8(this.size, v, true);
    }
    this.size += FieldSizes[FieldTypes.TypeUint8];
  }

  writeUint16(buf, v) {
    this.size = alignOffsetToType(this.size, FieldTypes.TypeUint16);
    if (buf) {
      new DataView(buf.buffer, buf.byteOffset).setUint16(this.size, v, true);
    }
    this.size += FieldSizes[FieldTypes.TypeUint16];
  }

  writeUint32(buf, v) {
    this.size = alignOffsetToType(this.size, FieldTypes.TypeUint32);
    if (buf) {
      new DataView(buf.buffer, buf.byteOffset).setUint32(this.size, v, true);
    }
    this.size += FieldSizes[FieldTypes.TypeUint32];
  }

  writeUint64(buf, v) {
    this.size = alignOffsetToType(this.size, FieldTypes.TypeUint64);
    if (buf) {
      new DataView(buf.buffer, buf.byteOffset).setBigUint64(this.size, v, true);
    }
    this.size += FieldSizes[FieldTypes.TypeUint64];
  }

  writeBytes(buf, v) {
    const dataView = (buf) ? new DataView(buf.buffer, buf.byteOffset) : undefined;
    this.size = alignOffsetToType(this.size, FieldTypes.TypeBytes);
    if (buf) {
      if (v) {
        dataView.setUint32(this.size, v.byteLength, true);
      } else {
        dataView.setUint32(this.size, 0, true);
      }
    }
    this.size += FieldSizes[FieldTypes.TypeBytes];
    this.size = alignDynamicFieldContentOffset(this.size, FieldTypes.TypeBytes);
    if (v) {
      if (buf) {
        buf.set(v, this.size);
      }
      this.size += v.byteLength;
    }
  }

  writeString(buf, v) {
    this.writeBytes(buf, getTextEncoder().encode(v));
  }

  writeUnionIndex(buf, unionIndex) {
    this.size = alignOffsetToType(this.size, FieldTypes.TypeUnion);
    if (buf) {
      new DataView(buf.buffer, buf.byteOffset).setUint16(this.size, unionIndex, true);
    }
    this.size += FieldSizes[FieldTypes.TypeUnion];
  }

  writeUint8Array(buf, v) {
    this.size = alignOffsetToType(this.size, FieldTypes.TypeUint8Array);
    if (buf) {
      new DataView(buf.buffer, buf.byteOffset).setUint32(this.size, v.length * FieldSizes[FieldTypes.TypeUint8], true);
    }
    this.size += FieldSizes[FieldTypes.TypeUint8Array];
    this.size = alignDynamicFieldContentOffset(this.size, FieldTypes.TypeUint8Array);
    for (const vv of v) {
      this.writeUint8(buf, vv);
    }
  }

  writeUint16Array(buf, v) {
    this.size = alignOffsetToType(this.size, FieldTypes.TypeUint16Array);
    if (buf) {
      new DataView(buf.buffer, buf.byteOffset).setUint32(this.size, v.length * FieldSizes[FieldTypes.TypeUint16], true);
    }
    this.size += FieldSizes[FieldTypes.TypeUint16Array];
    this.size = alignDynamicFieldContentOffset(this.size, FieldTypes.TypeUint16Array);
    for (const vv of v) {
      this.writeUint16(buf, vv);
    }
  }

  writeUint32Array(buf, v) {
    this.size = alignOffsetToType(this.size, FieldTypes.TypeUint32Array);
    if (buf) {
      new DataView(buf.buffer, buf.byteOffset).setUint32(this.size, v.length * FieldSizes[FieldTypes.TypeUint32], true);
    }
    this.size += FieldSizes[FieldTypes.TypeUint32Array];
    this.size = alignDynamicFieldContentOffset(this.size, FieldTypes.TypeUint32Array);
    for (const vv of v) {
      this.writeUint32(buf, vv);
    }
  }

  writeUint64Array(buf, v) {
    this.size = alignOffsetToType(this.size, FieldTypes.TypeUint64Array);
    if (buf) {
      new DataView(buf.buffer, buf.byteOffset).setUint32(this.size, v.length * FieldSizes[FieldTypes.TypeUint64], true);
    }
    this.size += FieldSizes[FieldTypes.TypeUint64Array];
    this.size = alignDynamicFieldContentOffset(this.size, FieldTypes.TypeUint64Array);
    for (const vv of v) {
      this.writeUint64(buf, vv);
    }
  }

  writeBytesArray(buf, v) {
    this.size = alignOffsetToType(this.size, FieldTypes.TypeBytesArray);
    const sizePlaceholderOffset = this.size;
    this.size += FieldSizes[FieldTypes.TypeBytesArray];
    this.size = alignDynamicFieldContentOffset(this.size, FieldTypes.TypeBytesArray);
    const contentSizeStartOffset = this.size;
    for (const vv of v) {
      this.writeBytes(buf, vv);
    }
    const contentSize = this.size - contentSizeStartOffset;
    if (buf) {
      new DataView(buf.buffer, buf.byteOffset).setUint32(sizePlaceholderOffset, contentSize, true);
    }
  }

  writeStringArray(buf, v) {
    this.size = alignOffsetToType(this.size, FieldTypes.TypeStringArray);
    const sizePlaceholderOffset = this.size;
    this.size += FieldSizes[FieldTypes.TypeStringArray];
    this.size = alignDynamicFieldContentOffset(this.size, FieldTypes.TypeStringArray);
    const contentSizeStartOffset = this.size;
    for (const vv of v) {
      this.writeString(buf, vv);
    }
    const contentSize = this.size - contentSizeStartOffset;
    if (buf) {
      new DataView(buf.buffer, buf.byteOffset).setUint32(sizePlaceholderOffset, contentSize, true);
    }
  }

  writeMessage(buf, v) {
    this.size = alignOffsetToType(this.size, FieldTypes.TypeMessage);
    const sizePlaceholderOffset = this.size;
    this.size += FieldSizes[FieldTypes.TypeMessage];
    this.size = alignDynamicFieldContentOffset(this.size, FieldTypes.TypeMessage);
    if (buf) {
      v.write(buf.subarray(this.size));
    } else {
      v.write(null);
    }
    const contentSize = v.getSize();
    this.size += contentSize;
    if (buf) {
      new DataView(buf.buffer, buf.byteOffset).setUint32(sizePlaceholderOffset, contentSize, true);
    }
  }

  writeMessageArray(buf, v) {
    this.size = alignOffsetToType(this.size, FieldTypes.TypeMessageArray);
    const sizePlaceholderOffset = this.size;
    this.size += FieldSizes[FieldTypes.TypeMessageArray];
    this.size = alignDynamicFieldContentOffset(this.size, FieldTypes.TypeMessageArray);
    const contentSizeStartOffset = this.size;
    for (const vv of v) {
      this.writeMessage(buf, vv);
    }
    const contentSize = this.size - contentSizeStartOffset;
    if (buf) {
      new DataView(buf.buffer, buf.byteOffset).setUint32(sizePlaceholderOffset, contentSize, true);
    }
  }

}